tests/validation/tests/v3_api/test_cert_rotation.py

```diff
@@ -0,0 +1,141 @@
+ALL_CERTS = ["kube-apiserver", "kube-controller-manager",
+             "kube-node", "kube-proxy", "kube-scheduler",
+             "kube-etcd", "kube-ca"]
+
+
+def test_rotate_all_certs():
+    changed = ALL_CERTS.copy()
+    changed.remove("kube-ca")
+    unchanged = ["kube-ca"]
+    rotate_and_compare(unchanged, changed)
+
+
+def test_rotate_kube_apiserver():
+    changed = ["kube-apiserver"]
+    unchanged = ALL_CERTS.copy()
+    unchanged.remove("kube-apiserver")
+    rotate_and_compare(unchanged, changed, "kube-apiserver")
+
+
+def test_rotate_kube_controller_manager():
+    changed = ["kube-controller-manager"]
+    unchanged = ALL_CERTS.copy()
+    unchanged.remove("kube-controller-manager")
+    rotate_and_compare(unchanged, changed, "kube-controller-manager")
+
+
+def test_rotate_kube_etcd():
+    changed = ["kube-etcd"]
+    unchanged = ALL_CERTS.copy()
+    unchanged.remove("kube-etcd")
+    rotate_and_compare(unchanged, changed, "etcd")
+
+
+def test_rotate_kube_node():
+    changed = ["kube-node"]
+    unchanged = ALL_CERTS.copy()
+    unchanged.remove("kube-node")
+    rotate_and_compare(unchanged, changed, "kubelet")
+
+
+def test_rotate_kube_proxy():
+    changed = ["kube-proxy"]
+    unchanged = ALL_CERTS.copy()
+    unchanged.remove("kube-proxy")
+    rotate_and_compare(unchanged, changed, "kube-proxy")
+
+
+def test_rotate_kube_scheduler():
+    changed = ["kube-scheduler"]
+    unchanged = ALL_CERTS.copy()
+    unchanged.remove("kube-scheduler")
+    rotate_and_compare(unchanged, changed, "kube-scheduler")
+
+
+def test_rotate_kube_ca():
+    changed = ALL_CERTS
+    unchanged = []
+    rotate_and_compare(unchanged, changed, "kube-ca")
+
+
+# Gets the certificate expiration date and cert name. Stores them in a dict.
+def get_certs():
+    certs = {}
+    client, cluster = get_user_client_and_cluster()
+    for key in cluster.certificatesExpiration:
+        if "kube-etcd" not in key:
+            certs[key] = parse_datetime(cluster.certificatesExpiration[key]
+                                        ["expirationDate"])
+
+    # Get etcd node certs from node IP
+    nodes = get_etcd_nodes(cluster)
+    for node in nodes:
+        if node["labels"]["node-role.kubernetes.io/etcd"] == "true":
+            ipKey = "kube-etcd-"+node["ipAddress"].replace(".", "-")
+            certs[ipKey] = parse_datetime(cluster.certificatesExpiration[ipKey]
+                                          ["expirationDate"])
+    return certs
+
+
+# Turn expiration string into datetime
+def parse_datetime(expiration_string):
+    return datetime.datetime.strptime(expiration_string, '%Y-%m-%dT%H:%M:%SZ')
+
+
+def compare_changed(certs2, time_now, changed):
+    if "kube-etcd" in changed:
+        for key in certs2:
+            if "kube-etcd" in key:
+                changed.append(key)
+        changed.remove("kube-etcd")
+    for i in changed:
+        if "kube-ca" in changed:
+            assert(certs2[i] > (time_now + datetime.timedelta(days=3647)))
```

Comment:
Validation of the expiry could be done as 365 * 10 days

Reply:
This has been changed to `assert(certs2[i] > (time_now.date() + datetime.timedelta(days=3650) - datetime.timedelta(minutes=10)))`.
The subtraction of 10 minutes is to allow for timing issues of `kube_ca` because the rotation takes more time than other rotations.

Reply:
Timing issue was from Jenkins side. Has been addressed.